libs/wingsdk/src/target-sim/function.inflight.ts

```diff
@@ -9,63 +13,176 @@ import {
-  private readonly sandbox: Sandbox;
+  private readonly maxWorkers: number;
+  private readonly workers = new Array<Sandbox>();
+  private createBundlePromise: Promise<void>;
 
   constructor(props: FunctionSchema["props"], context: ISimulatorContext) {
     if (props.sourceCodeLanguage !== "javascript") {
       throw new Error("Only JavaScript is supported");
     }
-    this.filename = path.resolve(context.simdir, props.sourceCodeFile);
+    this.originalFile = path.resolve(context.simdir, props.sourceCodeFile);
     this.env = props.environmentVariables ?? {};
     this.context = context;
     this.timeout = props.timeout;
-    this.sandbox = new Sandbox(this.filename, {
-      env: {
-        ...this.env,
-        WING_SIMULATOR_URL: this.context.serverUrl,
-      },
-      timeout: this.timeout,
-      log: (internal, _level, message) => {
-        this.context.addTrace({
-          data: { message },
-          type: internal ? TraceType.RESOURCE : TraceType.LOG,
-          sourcePath: this.context.resourcePath,
-          sourceType: FUNCTION_FQN,
-          timestamp: new Date().toISOString(),
-        });
-      },
-    });
+    this.maxWorkers = props.concurrency;
+
+    this.createBundlePromise = this.createBundle();
+  }
+
+  // Create a bundle with the code that will be run by all workers.
+  private async createBundle() {
+    const workdir = await mkdtemp(path.join(tmpdir(), "wing-bundles-"));
+
+    let contents = (await readFile(this.originalFile)).toString();
+
+    // log a warning if contents includes __dirname or __filename
+    if (contents.includes("__dirname") || contents.includes("__filename")) {
+      this.addTrace(
+        `Warning: __dirname and __filename cannot be used within bundled cloud functions. There may be unexpected behavior.`,
+        false
+      );
+    }
+
+    // wrap contents with a shim that handles the communication with the parent process
+    // we insert this shim before bundling to ensure source maps are generated correctly
+    contents = `
+"use strict";
+${contents}
+process.on("message", async (message) => {
+  const { fn, args } = message;
+  try {
+    const value = await exports[fn](...args);
+    process.send({ type: "resolve", value });
+  } catch (err) {
+    process.send({ type: "reject", reason: err });
+  }
+});
+`;
+    const wrappedPath = this.originalFile.replace(/\.js$/, ".sandbox.js");
+    writeFileSync(wrappedPath, contents); // async fsPromises.writeFile "flush" option is not available in Node 20
+    const bundle = createBundle(wrappedPath, [], workdir);
+    this.bundledFile = bundle.entrypointPath;
+
+    if (process.env.DEBUG) {
+      const fileStats = await stat(this.originalFile);
+      this.addTrace(`Bundled code (${fileStats.size} bytes).`);
+    }
   }
 
   public async init(): Promise<FunctionAttributes> {
     return {};
   }
 
   public async cleanup(): Promise<void> {
-    await this.sandbox.cleanup();
+    await this.createBundlePromise;
```

Comment:
Is this `await` needed?

Reply:
yeah looks off

Reply:
Yeah, I can see how this looks odd. I think there's a good reason in this scenario, but let me know what you think.

The reason I added this was to fix a rare bug that was causing `wing test` to crash sometimes. Here's what was happening: 
- `wing test` is executed with cleanup enabled (the default option). The app being tested has a cloud.Function which is not invoked during the test.
- The simulator starts, runs the tests, and stops shortly thereafter.
- When the simulator is cleaning up each resource, `cleanup()` is called on the cloud.Function.
- Since `cloud.Function` starts bundling work asynchronously during the constructor, it's possible the bundling won't be finished when the resource is being cleaned up.
- All resources are stopped, the simulator stops, and `wing test` cleans up files like the `.wing` directory etc.
- The bundling work is eventually executed, but it can't find a file, and a gross error is printed to the terminal.

There isn't really a way to cancel promises you've created in Node.js, so the cleanest solution seemed like just to wait for the work to finish in all code paths.

Anyways -- I'll add a comment to the code :-)